core/templates/components/graph-services/graph-layout.service.spec.ts

```diff
@@ -0,0 +1,1207 @@
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      }
+    };
+    let graphWidth = sgls.getGraphWidth(
+      AppConstants.MAX_NODES_PER_ROW, AppConstants.MAX_NODE_LABEL_LENGTH);
+    let graphHeight = sgls.getGraphHeight(nodeData);
+    let nodeDataWithPositionValueInPixel = sgls.modifyPositionValues(
+      nodeData, graphWidth, graphHeight);
+    let expectedGraphBoundaries = {
+      bottom: 245.8,
+      left: 40.675000000000004,
+      right: 589.325,
+      top: 34.2
+    };
+
+    let expectedWidth = expectedGraphBoundaries.bottom +
+      expectedGraphBoundaries.top;
+
+    expect(expectedWidth).toBeLessThanOrEqual(graphWidth);
+    expect(sgls.getGraphBoundaries(nodeDataWithPositionValueInPixel))
+      .toEqual(expectedGraphBoundaries);
+  });
+
+  it('should return graph boundaries with height equal to' +
+  ' the graph height', () => {
+    let nodeData = {
+      State1: {
+        depth: 1,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.42000000000000004,
+        xLabel: 0.1625,
+        yLabel: 0.5,
+        id: 'State1',
+        label: 'State1',
+        height: 0.16,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      Introduction: {
+        depth: 0,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.15333333333333335,
+        xLabel: 0.1625,
+        yLabel: 0.23333333333333334,
+        id: 'Introduction',
+        label: 'Introduction',
+        height: 0.16,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      End: {
+        depth: 2,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.6866666666666666,
+        xLabel: 0.1625,
+        yLabel: 0.7666666666666666,
+        id: 'End',
+        label: 'End',
+        height: 0.16,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State2: {
+        depth: 1,
+        offset: 1,
+        reachable: true,
+        x0: 0.29750000000000004,
+        y0: 0.42000000000000004,
+        xLabel: 0.3875,
+        yLabel: 0.5,
+        id: 'State2',
+        label: 'State2',
+        height: 0.16,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      }
+    };
+    let graphWidth = sgls.getGraphWidth(
+      AppConstants.MAX_NODES_PER_ROW, AppConstants.MAX_NODE_LABEL_LENGTH);
+    let graphHeight = sgls.getGraphHeight(nodeData);
+    let nodeDataWithPositionValueInPixel = sgls.modifyPositionValues(
+      nodeData, graphWidth, graphHeight);
+
+    let expectedGraphBoundaries = {
+      bottom: 182.79999999999998,
+      left: 40.675000000000004,
+      right: 305.82500000000005,
+      top: 27.200000000000003
+    };
+
+    let expectedHeight = expectedGraphBoundaries.bottom +
+      expectedGraphBoundaries.top;
+
+    expect(expectedHeight).toBeLessThanOrEqual(graphHeight);
```

Comment:
Why "bottom + top", and why <= when the description says equal?

More explanation of what is going on would be helpful.

Reply:
Added explanation [here](https://github.com/oppia/oppia/pull/12946#discussion_r644413438)